x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/validators/esql_query_validator_factory.ts

```diff
@@ -0,0 +1,180 @@
+    const [{ value }] = args;
+    const esqlQuery = value.query.query as string;
+
+    if (isEmpty(esqlQuery)) {
+      return;
+    }
+
+    try {
+      const services = KibanaServices.get();
+      const { isEsqlQueryAggregating, hasMetadataOperator, errors } = parseEsqlQuery(esqlQuery);
+
+      // Check if there are any syntax errors
+      if (errors.length) {
+        return constructSyntaxError(new Error(errors[0].message));
+      }
+
+      // non-aggregating query which does not have metadata, is not a valid one
+      if (!isEsqlQueryAggregating && !hasMetadataOperator) {
+        return {
+          code: ESQL_ERROR_CODES.ERR_MISSING_ID_FIELD_FROM_RESULT,
+          message: ESQL_VALIDATION_MISSING_METADATA_OPERATOR_IN_QUERY_ERROR,
+        };
+      }
+
+      const columns = await queryClient.fetchQuery({
+        queryKey: [esqlQuery.trim()],
+        queryFn: () =>
+          getESQLQueryColumns({
+            esqlQuery,
+            search: services.data.search.search,
+          }),
+      });
+
+      // for non-aggregating query, we want to disable queries w/o _id property returned in response
+      if (!isEsqlQueryAggregating && !columns.some(({ id }) => '_id' === id)) {
+        return {
+          code: ESQL_ERROR_CODES.ERR_MISSING_ID_FIELD_FROM_RESULT,
+          message: ESQL_VALIDATION_MISSING_ID_FIELD_IN_QUERY_ERROR,
+        };
+      }
+    } catch (error) {
+      return constructValidationError(error);
+    }
+  };
+}
+
+/**
+ * check if esql query valid for Security rule:
+ * - if it's non aggregation query it must have metadata operator
+ */
+function parseEsqlQuery(query: string) {
+  const { root, errors } = parse(query);
+  const isEsqlQueryAggregating = isAggregatingQuery(root);
+
+  return {
+    errors,
+    isEsqlQueryAggregating,
+    hasMetadataOperator: computeHasMetadataOperator(root),
+  };
+}
+
+/**
+ * checks whether query has metadata _id operator
+ */
+function computeHasMetadataOperator(astExpression: ESQLAstQueryExpression): boolean {
+  // Check whether the `from` command has `metadata` operator
+  const metadataOption = getMetadataOption(astExpression);
+  if (!metadataOption) {
+    return false;
+  }
+
+  // Check whether the `metadata` operator has `_id` argument
+  const idColumnItem = metadataOption.args.find(
+    (fromArg) => isColumnItem(fromArg) && fromArg.name === '_id'
+  );
+  if (!idColumnItem) {
+    return false;
+  }
+
+  return true;
+}
+
+function getMetadataOption(astExpression: ESQLAstQueryExpression): ESQLCommandOption | undefined {
+  const fromCommand = astExpression.commands.find((x) => x.name === 'from');
+
+  if (!fromCommand?.args) {
+    return undefined;
+  }
+
+  // Check whether the `from` command has `metadata` operator
+  for (const fromArg of fromCommand.args) {
+    if (isOptionItem(fromArg) && fromArg.name === 'metadata') {
+      return fromArg;
+    }
+  }
+
+  return undefined;
+}
+
+function constructSyntaxError(error: Error): ValidationError {
```

Comment:
Function expressions have better readability than arrow functions when defined at the file's top level. Of course it's debatable. In my experience I heard more complains from engineers checking upon code with arrow functions at the top level regarding readability. Since then I prefer using function expressions for better readability.